src/supervisor.rs

```diff
@@ -0,0 +1,614 @@
+                        //
+                        StartResult::SuccessfulStart {
+                            started_workers,
+                            pending_count,
+                        }
+                    }
+                    // when the child_spec failed to start
+                    Err((failed_worker_spec, start_error0)) => {
+                        let child_runtime_name =
+                            format!("{}/{}", sup_runtime_name, failed_worker_spec.name);
+                        let start_error = Arc::new(start_error0);
+                        // send event notification indicating that the worker failed to start
+                        ev_notifier
+                            .worker_start_failed(child_runtime_name, start_error.clone())
+                            .await;
+                        // transform current start to FailedStart
+                        StartResult::FailedStart {
+                            failed_worker_spec,
+                            start_error,
+                            // store the already started workers so that we can
+                            // later terminate them
+                            prev_started_workers: started_workers,
+                            // use the updated pending_count entry to get the
+                            // remaining pending_workers vec size
+                            pending_workers: Vec::with_capacity(pending_count),
+                        }
+                    }
+                }
+            }
+            // If we already have a failed start, we just append entries to the
+            // pending_workers vec
+            StartResult::FailedStart {
+                failed_worker_spec,
+                start_error,
+                prev_started_workers,
+                mut pending_workers,
+            } => {
+                pending_workers.push(child_spec);
+                StartResult::FailedStart {
+                    failed_worker_spec,
+                    start_error,
+                    prev_started_workers,
+                    pending_workers,
+                }
+            }
+        }
+    }
+}
+
+async fn start_children<'a, 'b>(
+    ctx: &'a Context,
+    mut ev_notifier: EventNotifier,
+    sup_runtime_name: &str,
+    sup_order: &'b Order,
+    // TODO: receive error notifier parameter
+    children_specs0: Vec<worker::Spec>,
+) -> Result<Vec<Worker>, (Vec<worker::Spec>, StartError)> {
+    let mut start_result = StartResult::new(children_specs0.len());
+    let pending_children: Box<dyn Iterator<Item = worker::Spec>> =
+        ordered_children_on_start(sup_order, children_specs0);
+
+    for failed_child_spec in pending_children {
+        start_result = start_result
+            .start_child(
+                ctx,
+                ev_notifier.clone(),
+                sup_runtime_name,
+                failed_child_spec,
+            )
+            .await;
+    }
```

Comment:
I think this iteration is part of what bothers me about the `StartResult` type. Instead of just unconditionally looping and the state type deciding what to do, it would make more sense to me to stop the iteration and then insert all the remaining pending children into the error.  You can do that with this same structure with something like:
```suggestion
    let mut error = Ok(());
    // Note the addition of `&mut` to not move the iterator itself.
    for failed_child_spec in &mut pending_children {
        let result = start_child(
                ctx,
                ev_notifier.clone(),
                sup_runtime_name,
                failed_child_spec,
            )
            .await;
            if let Err(start_error) = result {
                error = Err(start_error);
                break;
            }
    }
    // We can continue using `pending_children` here because we didn't move it into the for loop
    let pending_children: Vec<_> = pending_children.collect();
```

Reply:
I'm not really sure that this would end up being cleaner or simpler than what you have now, but I wanted to put it out there as a possibility since the current design is bothering some part of me.

Reply:
It does bother me too, I'm thinking, we would have `pending_children` have all the entries again, not the vector up to the point it call the `break` statement right? Some slicing needs to be done here with an index that I update on the loop.

Reply:
I'm not quite following you here. Are you asking whether the child entries will be removed from the `children_specs0` `Vec`? The answer to that is that they will. There is some sort of subtle ownership and borrowing happening here.  There are three types of iterators, corresponding to the types of ownership and borrowing, owned, ref, and mut ref.

```rust
let pending_children: Box<dyn Iterator<Item = worker::Spec>> =
        ordered_children_on_start(sup_order, children_specs0);
```
This line creates an owned iterator by moving the `children_specs0` vector into the iterator. That new `Box<dyn Iterator<Item = worker::Spec>>` will yield values of type `worker::Spec` no matter how we call it.

```rust
    for failed_child_spec in &mut pending_children {
        ...
    }
```
this usage of `&mut` _only applies to the iterator, and allows us to keep using the iterator after the loop is finished. This isn't useful normally, but because we might break from the loop early, in the error case that would mean the iterator has not finished iteration, and we can than capture the uniterated items into a new collection (with the `pending_children.collect();`).

Making this change you would probably want to substantially change the body of the loop and the start_child function to reflect the new type structure out here, I would move the accumulation of started `Child` structs into the loop, and just have `start_child` return those, wrapped in a `Result`. Then we don't need the `StartResult` type at all I think.

Reply:
> and allows us to keep using the iterator after the loop is finished

What I did not internalized, was that the iterator is stateful, so if we break the loop, next iteration won't start at the beginning of the collection, but where we were at at the end of the break.

I did a simplification of the code, I'm happier with the current solution. Thanks for the feedback and coaching